src/instamatic/microscope/typing.py

```diff
@@ -0,0 +1,13 @@
+from __future__ import annotations
+
+from typing import NamedTuple
+
+from instamatic.typing import float_deg, int_nm
+
+
+class StagePositionTuple(NamedTuple):
+    x: int_nm
+    y: int_nm
+    z: int_nm
+    a: float_deg
+    b: float_deg
```

Comment:
While you are at it, consider making this a dataclass, making it frozen for speed: `@dataclass(frozen=True)`. I think this will also help with the type specification in the other bits of the code, i.e. `def GetStagePositio(...) -> StagePosition`.

```suggestion
@dataclass(frozen=True)
class StagePosition:
    x: int_nm
    y: int_nm
    z: int_nm
    a: float_deg
    b: float_deg
```

Reply:
@stefsmeets Addressing individual issues:

- I do like your suggestion to add `Optional` for typing wherever it is applicable. I will do it in a separate commit to make sure all imports are in place.
- I just checked some files with `mypy` and currently the lack of `Optional` is the only issue found, `Annotated` are understood correctly.
- I do not have experience with `mypy` tests; this PR has, however, started from my trying to "enforce" the types using a wrapper class rather than just hinting at units – once I got it semi-functional I realized it is opaque and restrictive, hence I decided to go with very explicit type hints rather than opaque wrappers.
- Regarding `StagePositionTuple`, I thought about using it as output of individual interfaces but thought this might be too big of a difference... I think it would fit better though, so I will suggest adding it.
- I am strongly opposed to switching from `NamedTuple` to `@dataclass` here though: dataclass cannot be directly iterated or indexed, so returning a dataclass will be slower and will require a lot of code changes in applications IMO. Do not get me wrong, I do love dataclasses, but I think their strength lie elsewhere :).

Reply:
> I think this will also help with the type specification in the other bits of the code, i.e. `def GetStagePositio(...) -> StagePosition`.

In `[f584e5e](https://github.com/instamatic-dev/instamatic/pull/109/commits/f584e5efd560ce548714b679bbd09bf7a2ac9df0)`, I suggested now a change of the implementation of `StagePositionTuple` in all interfaces so that they return straight up an instance of `StagePositionTuple`. Is this what you had in mind? I didn't do that originally because I did not want the implementations to use any specific types. This is not obligatory since a normal `tuple` was converted to `NamedTuple` later on anyway, and should have no effect on performance since the cost of initializing `tuple` and `NamedTuple` is basically the same.

Reply:
Fair enough. From my experience there is not much in it in terms of performance between dataclasses and namedtuples. But I forgot that we still do tuple unpacking in many places (imo a major footgun).